REST_AUTHZ_FOR_PLUGINS.md

```diff
@@ -0,0 +1,134 @@
+Next we get the unique name and actionNames associated with that route and evaluate these against existing `cluster_permissions` across all roles of the requesting user. If the authorization check succeeds, the request chain proceeds as normal. If it fails, a 401 response is returned to the user.
+
+NOTE:
+1. The action names defined in roles must exactly match the names of registered routes, or else, the request would be deemed unauthorized.
+2. This check will not be implemented for plugins who do not use NamedRoutes.
+
+
+
+### How to translate an existing Route to be a NamedRoute?
+
+Here is a sample of an existing route converted to a named route:
+Before:
+```
+public List<Route> routes() {
+    return ImmutableList.of(
+            new Route(GET, "/uri")
+        );
+}
+```
+With new scheme:
+```
+public List<NamedRoute> routes() {
+    return ImmutableList.of(
+            new NamedRoute.Builder().method(GET).path("/uri").uniqueName("plugin:uri").actionNames(Set.of("cluster:admin/opensearch/plugin/uri")).build()
+        );
+}
+```
+
+`actionNames()` are optional. They correspond to any current actions defined as permissions in roles.
+Ensure that these name-to-route mappings are easily accessible to the cluster admins to allow granting access to these APIs.
+
+### How does authorization in the REST Layer work?
+
+We will continue on the above example of translating `/uri` from Route to NamedRoute.
+
+Consider these roles are defined in the cluster:
+```yaml
+plugin_role:
+  reserved: true
+  cluster_permissions:
+    - 'plugin:uri'
+
+plugin_role_legacy:
+  reserved: true
+  cluster_permissions:
+    - 'cluster:admin/opensearch/plugin/uri'
+```
+
+Successful authz scenarios for a user:
+1. The user is mapped either to `plugin_role` OR `plugin_role_legacy`.
+2. The user is mapped to both of these roles.
+3. The user is mapped to `plugin_role` even if no `actionNames()` were registered for this route.
+
+Unsuccessful authz scenarios for a user:
+1. The user is not mapped any roles.
+2. The user is mapped to a different role which doesn't grant the cluster permissions: `plugin:uri` OR `cluster:admin/opensearch/plugin/uri`/
+3. The user is mapped to a role `plugin_role_other` which has a typo in action name, i.e.`plugin:uuri`.
+
+
+### Sample API in Security Plugin
+
+As part of this effort a new uri `GET /whoamiprotected` was introduced as a NamedRoute version of `GET /whoami`. Here is how you can test it:
+
+#### roles.yml
+```yaml
+who_am_i_role:
+  reserved: true
+  cluster_permissions:
+    - 'security:whoamiprotected'
+
+who_am_i_role_legacy:
+  reserved: true
+  cluster_permissions:
+    - 'cluster:admin/opendistro_security/whoamiprotected'
+
+who_am_i_role_no_perm:
+  reserved: true
+  cluster_permissions:
+    - 'some_invalid_perm'
+
+```
+
+#### internal_users.yml
+```yaml
+who_am_i-user:
```

Comment:
nit: mix of dashes and underscores

Reply:
I kept this on purpose, to differentiate between a role and a user since they have same prefix.